lib/views/addToCollection.js

```diff
@@ -0,0 +1,119 @@
+const loadTemplate = require('../loadTemplate')
+const pug = require('pug')
+const config = require('../config')
+const sparql = require('../sparql/sparql')
+const getOwnedBy = require('../query/ownedBy')
+
+module.exports = function (req, res) {
+  req.setTimeout(0) // no timeout
+
+  function addToCollectionsForm (req, res) {
+    var collectionQuery = 'PREFIX dcterms: <http://purl.org/dc/terms/> PREFIX sbol2: <http://sbols.org/v2#> SELECT ?object ?name WHERE { ?object a sbol2:Collection . OPTIONAL { ?object dcterms:title ?name . }}'
+
+    function sortByNames (a, b) {
+      if (a.name < b.name) {
+        return -1
+      } else {
+        return 1
+      }
+    }
+
+    return sparql.queryJson(collectionQuery, req.user.graphUri).then((collections) => {
+      collections.forEach((result) => {
+        result.uri = result.object
+        result.name = result.name ? result.name : result.uri.toString()
+        delete result.object
+      })
+      collections.sort(sortByNames)
+
+      const locals = {
+        config: config.get(),
+        section: 'addToCollection',
+        user: req.user,
+        collections: collections,
+        addToCollectionUrl: req.url,
+        errors: {}
+      }
+      res.send(pug.renderFile('templates/views/addToCollection.jade', locals))
+    })
+  }
+
+  if (req.method === 'POST') {
+    const uri = req.body.collections.toString() // .replace(config.get('databasePrefix'),config.get('instanceUrl'))
+    const returnUrl = req.url.toString().replace('/addToCollection', '')
+    var memberUri = returnUrl.replace(config.get('instanceUrl'), config.get('databasePrefix'))
+
+    if (!memberUri) {
+      if (!req.accepts('text/html')) {
+        return res.status(400).type('text/plain').send('Must provide URI of member to add')
+      } else {
+        const locals = {
+          config: config.get(),
+          section: 'errors',
+          user: req.user,
+          errors: [ 'Must provide URI of member to add' ]
+        }
+        return res.status(400).send(pug.renderFile('templates/views/errors/errors.jade', locals))
+      }
+    }
+
+    if (memberUri.startsWith('/public/')) {
+      memberUri = memberUri.replace('/public/', config.get('databasePrefix') + 'public/')
+    }
+    if (memberUri.startsWith('/user/')) {
+      memberUri = memberUri.replace('/user/', config.get('databasePrefix') + 'user/')
+    }
+
+    var templateParams = {
+      uri: uri,
+      memberUri: memberUri
+    }
+
+    var addQuery = loadTemplate('sparql/addMembership.sparql', templateParams)
+
+    console.log(addQuery)
+
+    return getOwnedBy(uri, req.user.graphUri).then((ownedBy) => {
```

Comment:
Can we check ownedBy earlier? I think that makes more sense to read rather than interrupting the process of adding.

Reply:
Not sure where you have in mind.  There is not actual code earlier, since much of the code before is just a function for the form.